gtsam/hybrid/HybridConditional.h

```diff
@@ -0,0 +1,135 @@
+#include <string>
+#include <typeinfo>
+#include <vector>
+
+namespace gtsam {
+
+class HybridFactorGraph;
+
+/**
+ * Hybrid Conditional Density
+ *
+ * As a type-erased variant of:
+ * - DiscreteConditional
+ * - GaussianConditional
+ * - GaussianMixtureConditional
+ *
+ * The reason why this is important is that `Conditional<T>` is a CRTP class.
+ * CRTP is static polymorphism such that all CRTP classes, while bearing the
+ * same name, are different classes not sharing a vtable. This prevents them
+ * from being contained in any container, and thus it is impossible to
+ * dynamically cast between them. A better option, as illustrated here, is
+ * treating them as an implementation detail - such that the hybrid mechanism
+ * does not know what is inside the HybridConditional. This prevents us from
+ * having diamond inheritances, and neutralized the need to change other
+ * components of GTSAM to make hybrid elimination work.
+ *
+ * A great reference to the type-erasure pattern is Edurado Madrid's CppCon
+ * talk.
+ */
+class GTSAM_EXPORT HybridConditional
+    : public HybridFactor,
+      public Conditional<HybridFactor, HybridConditional> {
+ public:
+  // typedefs needed to play nice with gtsam
+  typedef HybridConditional This;              ///< Typedef to this class
+  typedef boost::shared_ptr<This> shared_ptr;  ///< shared_ptr to this class
+  typedef HybridFactor BaseFactor;  ///< Typedef to our factor base class
+  typedef Conditional<BaseFactor, This>
+      BaseConditional;  ///< Typedef to our conditional base class
+
+ protected:
+  // Type-erased pointer to the inner type
+  boost::shared_ptr<Factor> inner;
+
+ public:
+  /// @name Standard Constructors
+  /// @{
+
+  /// Default constructor needed for serialization.
+  HybridConditional() = default;
+
+  HybridConditional(const KeyVector& continuousKeys,
+                    const DiscreteKeys& discreteKeys, size_t nFrontals)
+      : BaseFactor(continuousKeys, discreteKeys), BaseConditional(nFrontals) {}
+
+  HybridConditional(const KeyVector& continuousFrontals,
+                    const DiscreteKeys& discreteFrontals,
+                    const KeyVector& continuousParents,
+                    const DiscreteKeys& discreteParents);
+
+  HybridConditional(
+      boost::shared_ptr<GaussianConditional> continuousConditional);
+
+  HybridConditional(boost::shared_ptr<DiscreteConditional> discreteConditional);
+
+  HybridConditional(
+      boost::shared_ptr<GaussianMixtureConditional> gaussianMixture);
+
+  GaussianMixtureConditional::shared_ptr asMixture() {
+    if (!isHybrid_) throw std::invalid_argument("Not a mixture");
+    return boost::static_pointer_cast<GaussianMixtureConditional>(inner);
+  }
+
+  DiscreteConditional::shared_ptr asDiscreteConditional() {
+    if (!isDiscrete_) throw std::invalid_argument("Not a discrete conditional");
+    return boost::static_pointer_cast<DiscreteConditional>(inner);
+  }
+
+  boost::shared_ptr<Factor> getInner() { return inner; }
```

Comment:
public?

Reply:
Moved to protected scoped.

Reply:
Nvm the method has to be public so it can be called by the factor graph.

Reply:
What I am doing is converting the variable from `inner` to `inner_` and adding the `inner()` method to all the classes so we're functional.